pkg/codegen/hcl2/model/type_object.go

```diff
@@ -218,20 +227,35 @@ func (t *ObjectType) conversionFrom(src Type, unifying bool) ConversionKind {
 }
 
 func (t *ObjectType) String() string {
-	if t.s == "" {
-		var properties []string
-		for k, v := range t.Properties {
-			properties = append(properties, fmt.Sprintf("%s = %v", k, v))
-		}
-		sort.Strings(properties)
+	return t.string(nil)
+}
+
+func (t *ObjectType) string(seen map[Type]struct{}) string {
+	if t.s != "" {
+		return t.s
+	}
 
-		annotations := ""
-		if len(t.Annotations) != 0 {
-			annotations = fmt.Sprintf(", annotated(%p)", t)
+	if seen != nil {
+		if _, ok := seen[t]; ok {
+			return "..."
```

Comment:
Can it be the case that t.s is not set but the object type exists in `seen`? Is that a panic worthy event? What does `...` actually mean in hcl2?

Reply:
> Can it be the case that t.s is not set but the object type exists in seen?

Yes - that's when the same type is the "child" of itself - as seen in the test https://github.com/pulumi/pulumi/pull/6250/files#diff-eff2ea294a07fa6b0e4ca9b4e2766d60c1e923cad0771d3deb9514d9b3a4036dR693

> What does ... actually mean in hcl2?

It means nothing in HCL. We aren't producing HCL here, we are producing a text representation of a type.